### PR TITLE
FIX1958: Error code 400 when input trace is too large(WIP)

### LINKED ISSUE #1958: Fix distributor http status codes
**Describe the bug**
As reported by the community if the ingester returns trace too large then the returned http error is 500. This error suggests to the otel collector to retry however it should return 400 which suggests to not retry. Currently the GRPC status codes correctly suggest retrying or not.

Review the status codes for various errors and return 400s where appropriate.

### PR DESCRIPTION
**Pull request by mghildiy**
**Return status code 400 when input trace is too large**:

**Currently, status code returned for traces larger than permissible size is 500, which makes client to retry. To avoid this, return status code 400 instead.**:
Fixes #1958

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/ingester/instance.go [hunk 1]
```diff
@@ -186,7 +186,7 @@ func (i *instance) push(ctx context.Context, id, traceBytes []byte) error {
 		prevSize := int(i.traceSizes[tkn])
 		reqSize := len(traceBytes)
 		if prevSize+reqSize > maxBytes {
-			return status.Errorf(codes.FailedPrecondition, (newTraceTooLargeError(id, i.instanceID, maxBytes, reqSize).Error()))
+			return status.Errorf(codes.InvalidArgument, (newTraceTooLargeError(id, i.instanceID, maxBytes, reqSize).Error()))
```
Comment (joe-elliott) [inline modules/ingester/instance.go:189]: have you tested this. does this really cause the distributor to return a 400?

Comment (mghildiy) [reply modules/ingester/instance.go:189]: This is WIP.

Comment (mghildiy) [reply modules/ingester/instance.go:189]: How do I test it? I have created a small spring boot app, and configured it with otel to send traces to tempo instance. 
I face 2 issues:

1. I am not able to configure max_bytes_per_trace for overrides. When I do it, tempo doesnt start, though its a valid config parameter as per official documentation.

```
overrides:
  per_tenant_override_config: /etc/overrides.yaml
```

In overrides.yaml, I have:
```
overrides:
  max_bytes_per_trace: 500
```

2. As a workaround, I hardcode  this config in instance.go. But this stops even other requests from being accepted(I see some requests with request bytes size around 10064 at repeated intervals. I think they are compaction related activity)

Comment (mghildiy) [reply modules/ingester/instance.go:189]: In collector logs I see:

> 023-07-12 19:05:03 2023-07-12T13:35:03.418Zerrorexporterhelper/queued_retry.go:391Exporting failed. The error is not retryable. Dropping data.{"kind": "exporter", "data_type": "traces", "name": "otlp", "error": "Permanent error: rpc error: code = InvalidArgument desc = TRACE_TOO_LARGE: max size of trace (500) exceeded while adding 5024 bytes to trace 64aeac0154073ef25bea20d464cc57d3 for tenant single-tenant", "dropped_items": 4}

Seems good to me.

Comment (mghildiy) [reply modules/ingester/instance.go:189]: Is there a way to turn off compaction?

Comment (joe-elliott) [reply modules/ingester/instance.go:189]: Perhaps try something like:

```
compactor:
  compaction:
    max_block_bytes: 1
```

Comment (mghildiy) [reply modules/ingester/instance.go:189]: In collector logs, I see error message containing 'The error is not retryable. Dropping data' no matter which code I use in ingester,  FailedPrecondition OR InvalidArgument.

**Pull request closed without merging**